docs/src/test/java/docs/http/javadsl/CustomHeaderExampleTest.java

```diff
@@ -0,0 +1,118 @@
+package docs.http.javadsl;
+
+import akka.http.javadsl.model.HttpHeader;
+import akka.http.javadsl.model.HttpRequest;
+import akka.http.javadsl.model.StatusCodes;
+import akka.http.javadsl.model.headers.ModeledCustomHeader;
+import akka.http.javadsl.model.headers.ModeledCustomHeaderCompanion;
+import akka.http.javadsl.model.headers.RawHeader;
+import akka.http.javadsl.server.Route;
+import akka.http.javadsl.testkit.JUnitRouteTest;
+import akka.japi.JavaPartialFunction;
+import org.junit.Test;
+import scala.PartialFunction;
+
+import java.util.Optional;
+
+import static org.junit.Assert.*;
+
+public class CustomHeaderExampleTest extends JUnitRouteTest {
+  //#modeled-api-key-custom-header
+  public static class ApiTokenHeader extends ModeledCustomHeader {
+
+    ApiTokenHeader(String name, String value) {
+      super(name, value);
+    }
+
+    public boolean renderInResponses() {
+      return false;
+    }
+
+    public boolean renderInRequests() {
+      return false;
+    }
+
+  }
+
+  static class ApiTokenHeaderComp extends ModeledCustomHeaderCompanion<ApiTokenHeader> {
+
+    public String name() {
+      return "apiKey";
+    }
+
+    @Override
+    public ApiTokenHeader parse(String value) {
+      return new ApiTokenHeader(name(), value);
+    }
+
+  }
+  //#modeled-api-key-custom-header
+
+  @Test
+  public void showCreation() {
+    //#conversion-creation-custom-header
+    final ApiTokenHeaderComp apiTokenHeaderComp = new ApiTokenHeaderComp();
+    final ApiTokenHeader token = apiTokenHeaderComp.create("token");
+    assertEquals("token", token.value());
+
+    final HttpHeader header = apiTokenHeaderComp.create("token");
+    assertEquals("apiKey", header.name());
+    assertEquals("token", header.value());
+
+    final Optional<ApiTokenHeader> fromRaw = apiTokenHeaderComp
+      .from(RawHeader.create("apiKey", "token"));
```

Comment:
not pattern matching, but looks a convenient way of transforming a `RawHeader` to your custom one.